File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int best_sum = 0, curr_sum = 0;
        for (int i = 1; i < prices.size(); i++) {
            curr_sum = std::max(0, curr_sum += prices[i] - prices[i - 1]);
            best_sum = std::max(curr_sum, best_sum);
        }
        return best_sum;
    }
};

File: top-interview-150/0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        max_profit = 0
        min_price = prices[0]
        
        for right in range(1, len(prices)):
            profit = prices[right] - min_price
            
            if profit > max_profit:
                max_profit = profit
                
            if prices[right] < min_price:
                min_price = prices[right]
                
        return max_profit
